db/version_set.cc

```diff
@@ -4974,19 +4974,27 @@ uint64_t VersionSet::ApproximateSize(Version* v, const FdWithKeyRange& f,
   } else {
     // "key" falls in the range for this table.  Add the
     // approximate offset of "key" within the table.
-    TableReader* table_reader_ptr;
-    InternalIterator* iter = v->cfd_->table_cache()->NewIterator(
-        ReadOptions(), v->env_options_, v->cfd_->internal_comparator(),
-        *f.file_metadata, nullptr /* range_del_agg */,
-        v->GetMutableCFOptions().prefix_extractor.get(), &table_reader_ptr,
-        /*file_read_hist=*/nullptr, caller,
-        /*arena=*/nullptr, /*skip_filters=*/false, /*level=*/-1,
-        /*smallest_compaction_key=*/nullptr,
-        /*largest_compaction_key=*/nullptr);
-    if (table_reader_ptr != nullptr) {
-      result = table_reader_ptr->ApproximateOffsetOf(key, caller);
-    }
-    delete iter;
+    auto& fd = f.file_metadata->fd;
+    TableCache* table_cache = v->cfd_->table_cache();
+    Cache::Handle* handle = nullptr;
+    TableReader* table_reader = fd.table_reader;
+    if (table_reader == nullptr && table_cache != nullptr) {
+      const bool for_compaction = (caller == TableReaderCaller::kCompaction);
+      Status s = table_cache->FindTable(
+          v->env_options_, v->cfd_->internal_comparator(), fd, &handle,
+          v->GetMutableCFOptions().prefix_extractor.get(), false /* no_io */,
+          !for_compaction /* record_read_stats */);
+      if (s.ok()) {
+        table_reader = table_cache->GetTableReaderFromHandle(handle);
+      }
+    }
+
+    if (table_reader != nullptr) {
+      result = table_reader->ApproximateOffsetOf(key, caller);
+    }
+    if (handle != nullptr && table_cache != nullptr) {
+      table_cache->ReleaseHandle(handle);
+    }
```

Comment:
Do we consider to move the whole logic to a new function in TableCache, like `TableCache::ApproximateOffsetOf()`?
I feel it cleaner to keep those table cache handle inside TableCache as much as possible.

Reply:
done